Improve block field functions by accessing sideband information internally

clemahieu opened the pull request:
This change moves the responsibility of determining if a field is in the block or in the block's sideband on to the block class.

Previously call sites would need to either inspect the block type or check for a sentinel value to determine where the block field information was stored. This complicates call sites, duplicates code, and could be error prone.

Direct block field access e.g. account/link/destination/balance are suffixed with _field and return an std::optional depending if the block type contains the requested field.

The replacement field functions check block type and determine if the value should be retrieved from the field or the block sideband.

The conversion for each field is done with two commits in order to ensure all call sites are inspected and updated to the new semantics.
- The first commit changes the return type of of the block field which allowed easy locating / updating of all the call sites through compiler type checking.
- The second commit renames the direct field access function and adds a helper function to pull the field value or the sideband value depending on block type.

clemahieu committed:
Directly use genesis_key.pub for obtaining the genesis account, rather than querying the genesis block.

clemahieu committed:
Convert boost::optional to std::optional for distributed work generation.

clemahieu committed:
Change block::account to return an std::optional if the block actually contains the field.

Fixes up call sites that checked against the old sentinel value 0.
Convert many call sites to using ledger::account(block&) when possible.

clemahieu committed:
ledger::account is moved to block::account and block::account which accesses the account field is renamed to block::account_field.

pwojcikdev commented on nano/node/bootstrap_ascending/service.cpp:
```diff
@@ -166,7 +166,7 @@ void nano::bootstrap_ascending::service::inspect (store::transaction const & tx,
 		case nano::block_status::gap_source:
 		{
-			const auto account = block.previous ().is_zero () ? block.account ().value () : ledger.account (tx, block.previous ()).value ();
+			const auto account = block.previous ().is_zero () ? block.account_field ().value () : ledger.account (tx, block.previous ()).value ();
```
This line looks a bit suspicious, it seems to rely on the fact that if both source and previous block is missing, then ledger processing code will return block_status::gap_previous. Otherwise it would crash. No need to do anything, just something that might bite later down the road, with subsequent refactorings.

clemahieu replied:
Yea it relies on the fact that if previous.is_zero (), then it must be an open or state block and therefore the account field is present.  Since the block failed to be inserted it won't have sideband at this point.

clemahieu committed:
Change block::balance to return an std::optional if the block actually contains the field.

clemahieu committed:
ledger::balance is moved to block::balance and block::balance which accesses the account field is renamed to block::balance_field.

clemahieu committed:
Change block::destination to return an std::optional if the block actually contains the field.

clemahieu committed:
ledger::destination is moved to block::destination_field and block::destination is created which accesses the the destination field for send blocks.

clemahieu committed:
Change block::source to return an std::optional if the block actually contains the field.

clemahieu committed:
ledger::source is moved to block::source_field and ledger::block_source is moved to block::source which accesses the source field for open/receive/state blocks that are receives.

clemahieu committed:
Change block::link to return an std::optional if the block actually contains the field.

clemahieu committed:
Moving ledger::is_send to block::is_send.

A specialised variant of is_send for state blocks that do not have sideband was added to dependent_block_visitor so the value can be calculated even if the block does not yet have sideband.

clemahieu committed:
Creating block::is_receive function to determine if a block is a receive/open/state block that's a receive

clemahieu committed:
Create block::is_change function which determines if a block is a change block or a state block with a zero link field.

clemahieu committed:
Renaming block::link to block::link_field. Usages that converted this number to an account number are replaced with calls to block::destination (). Usages that converted this number to a block_hash are replaced with calls to block::source, and usages that check this number for being an epoch directly access the link field.

clemahieu committed:
Change block::representative to return an std::optional if the block actually contains the field.

clemahieu committed:
Change block::previous to block::previous_field and return std::optional depending if the block has the field.

Add block::previous with backwards compatibility and a todo to fix up usages to not check for sentinel values.

pwojcikdev commented on nano/lib/blocks.cpp:
```diff
@@ -161,45 +208,118 @@ bool nano::block::has_sideband () const
+			release_assert (sideband ().details.is_send);
+			return link_field ().value ().as_account ();
+		default:
+			release_assert (false);
```
There is some inconsistency in the way unexpected behavior is handled, `.value ()` will throw an exception while `release_assert` will terminate the program. Maybe those accessors should be marked with `noexcept` to terminate instantly if they find the program in undefined state.

clemahieu replied:
Yea I can make that change. The intent with the value () calls is to terminate, at least in these inner functions.

clemahieu commented on nano/lib/blocks.hpp:
```diff
@@ -63,6 +52,35 @@ class block
+	virtual std::optional<nano::account> destination_field () const;
+	// Link field for state blocks
+	virtual std::optional<nano::link> link_field () const;
+	// Previous block if field exists and is non-zero
```
Comment needs to be corrected.

pwojcikdev commented on nano/node/active_transactions.cpp:
```diff
@@ -165,8 +165,10 @@ void nano::active_transactions::handle_confirmation (nano::store::read_transacti
 {
-	auto destination = block->link ().is_zero () ? block->destination () : block->link ().as_account ();
-	node.receive_confirmed (transaction, hash, destination);
+	if (block->sideband ().details.is_send)
```
Could be using `block->is_send()` helper

pwojcikdev commented on nano/node/json_handler.cpp:
```diff
@@ -3679,7 +3679,7 @@ void nano::json_handler::republish ()
 							{
 								hashes.push_back (previous);
-								source = node.ledger.block_source (transaction, *block_d);
+								source = block_d->source_field ().value_or (block_d->is_send () ? 0 : block_d->link_field ().value_or (0).as_block_hash ());
```
A short `node.ledger.block_source ( .. )` call changed into this relatively long conditional statement, is this intentional?

pwojcikdev commented on nano/node/json_handler.cpp:
```diff
@@ -3641,7 +3641,7 @@ void nano::json_handler::republish ()
 				if (sources != 0) // Republish source chain
 				{
-					nano::block_hash source (node.ledger.block_source (transaction, *block));
+					nano::block_hash source = block->source_field ().value_or (block->link_field ().value_or (0).as_block_hash ());
```
Another line that got more complicated after the refactor than before.
